code/modules/reagents/reagent_containers/food/drinks.dm

```diff
@@ -427,20 +427,197 @@ If you add a drink with no empty icon sprite, ensure it is flagged as NO_EMPTY_I
+		playsound(src.loc, /decl/sound_category/shaker_shaking, 50, 1)
 	src.add_fingerprint(user)
 	return
 
+/obj/item/reagent_containers/food/drinks/shaker/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/reagent_containers/food/drinks/shaker_cup))
+		if(cap)
+			to_chat(user, SPAN_WARNING("\The [src] already has \a [cap]."))
+			return TRUE
+		if(W.reagents.total_volume > 0)
+			var/obj/item/reagent_containers/food/drinks/shaker_cup/C = W
+			C.standard_pour_into(user, src)
+			return TRUE
+		if(!top)
+			to_chat(user, SPAN_WARNING("\The [src] lacks a top to fit \the [W] on."))
+			return TRUE
+		to_chat(user, SPAN_NOTICE("You put \the [W] onto \the [src]."))
+		user.drop_from_inventory(W, src)
+		flags ^= OPENCONTAINER
+		cap = W
+		playsound(src.loc, /decl/sound_category/shaker_lid_off, 50, 1)
+		update_icon()
+		return TRUE
+	if(istype(W, /obj/item/shaker_top))
+		if(top)
+			to_chat(user, SPAN_WARNING("\The [src] already has \a [top]."))
+			return TRUE
+		to_chat(user, SPAN_NOTICE("You fit \the [W] onto \the [src]."))
+		amount_per_transfer_from_this = 10
+		user.drop_from_inventory(W, src)
+		top = W
+		playsound(src.loc, /decl/sound_category/shaker_lid_off, 50, 1)
+		update_icon()
+		return TRUE
+	return ..()
+
+/obj/item/reagent_containers/food/drinks/shaker/on_pour()
+	if(!top)
+		playsound(src, 'sound/effects/pour_big.ogg', 50, 1)
+		return
+	return ..()
+
+/obj/item/reagent_containers/food/drinks/shaker/verb/toggle_twist()
+	set category = "Object"
+	set name = "Twist Cap"
+	set src in usr
+
+	if(!cap)
+		to_chat(usr, SPAN_WARNING("\The [src] doesn't have a cap!"))
+		return
+	twisted = !twisted
+	to_chat(usr, SPAN_NOTICE("You twist \the [cap] [twisted ? "closed for a tight seal" : "loose"]."))
+	update_icon()
+
+/obj/item/reagent_containers/food/drinks/shaker/verb/toggle_top()
+	set category = "Object"
+	set name = "Remove Top"
+	set src in usr
+
+	if(cap)
+		to_chat(usr, SPAN_WARNING("You must remove \the [cap] first!"))
+		return
+	if(!top)
+		to_chat(usr, SPAN_WARNING("\The [src] doesn't have a top to remove!"))
+		return
+	to_chat(usr, SPAN_NOTICE("You remove \the [src]'s [top]."))
+	amount_per_transfer_from_this = 120
+	usr.put_in_hands(top)
+	top = null
+	playsound(src.loc, /decl/sound_category/shaker_lid_off, 50, 1)
+	update_icon()
+
+/obj/item/shaker_top
+	name = "shaker top"
+	desc = "A metal shaker top with an in-built filter on the bottom."
+	desc_info = "When fitted on a shaker, you can Alt Click the shaker to change transfer amount of the shaker."
+	icon = 'icons/obj/shaker.dmi'
+	icon_state = "shaker_top"
+	item_state = "shaker_top"
+	contained_sprite = TRUE
+	drop_sound = 'sound/items/drop/bottle.ogg'
+	pickup_sound = null
+	center_of_mass = list("x" = 16, "y" = 16)
+
+/obj/item/shaker_top/attack_self(mob/user as mob)
+	return
+
+/obj/item/shaker_top/afterattack(atom/target, mob/user, proximity, params)
+	if(istype(target, /obj/item/reagent_containers/food/drinks/shaker))
+		return
+	return ..()
```

Comment:
This proc will never be called because you `return TRUE` in `/obj/item/reagent_containers/food/drinks/shaker/attackby(obj/item/W, mob/user)` when you're attacking it with `/obj/item/shaker_top`.

`return TRUE` in attackby means that the attack is resolved there and won't go on to call `afterattack`.

Reply:
Oooooooooh. I see, now. The very base code for the shaker was based from the ``flask/vacuumflask``, and since it called there and didn't seem to destroy anything, I just. Mindlessly copied it. I'll absolutely keep this in mind from now on. Thank you very much!